File: amsr/valence.py
# key: symbol, charge, number of !'s (hypervalent marker)
VALENCE = {
    ("H", 0, 0): 1,
    ("H", 1, 0): 0,
    ("H", -1, 0): 0,
    ("He", 0, 0): 0,
    ("Li", 1, 0): 0,
    ("Li", 0, 0): 1,
    ("Be", 2, 0): 0,
    ("B", 0, 0): 3,
    ("B", -1, 0): 4,
    ("C", 0, 0): 4,
    ("C", -1, 0): 3,
    ("C", 1, 0): 3,
    ("N", 0, 0): 3,
    ("N", 1, 0): 4,
    ("N", -1, 0): 2,
    ("O", 0, 0): 2,
    ("O", -1, 0): 1,
    ("O", -2, 0): 0,
    ("O", 1, 0): 3,
    ("F", 0, 0): 1,
    ("F", -1, 0): 0,
    ("F", 1, 0): 2,
    ("Na", 0, 0): 1,
    ("Na", 1, 0): 0,
    ("Mg", 0, 0): 2,
    ("Mg", 0, 1): 6,
    ("Mg", 1, 0): 1,
    ("Mg", 2, 0): 0,
    ("Al", 0, 0): 3,
    ("Al", 3, 0): 0,
    ("Al", -3, 0): 6,
    ("Xe", 0, 0): 0,
    ("Si", 0, 0): 4,
    ("Si", -1, 0): 5,
    ("Si", 4, 0): 0,
    ("P", 0, 0): 3,
    ("P", 0, 1): 5,
    ("P", 0, 2): 7,
    ("P", 1, 0): 4,
    ("P", -1, 1): 6,
    ("S", 0, 0): 2,
    ("S", 0, 1): 4,
    ("S", 0, 2): 6,
    ("S", 1, 0): 3,
    ("S", 1, 1): 5,
    ("S", -1, 0): 1,
    ("S", -1, 1): 3,
    ("S", -1, 2): 5,
    ("S", -2, 0): 0,
    ("Cl", 0, 0): 1,
    ("Cl", -1, 0): 0,
    ("Cl", 1, 0): 2,
    ("Cl", 2, 0): 3,
    ("Cl", 3, 0): 4,
    ("K", 0, 0): 1,
    ("K", 1, 0): 0,
    ("Ca", 0, 0): 2,
    ("Ca", 2, 0): 0,
    ("Fe", 3, 0): 0,
    ("Zn", 0, 0): 2,
    ("Zn", 1, 0): 1,
    ("Zn", 2, 0): 1,
    ("Zn", -2, 0): 2,
    ("Zn", -2, 1): 4,
    ("Ga", 3, 0): 0,
    ("As", 0, 0): 3,
    ("As", 0, 1): 5,
    ("As", 0, 2): 7,
    ("As", 1, 0): 4,
    ("As", -1, 1): 6,
    ("Se", 0, 0): 2,
    ("Se", 0, 1): 4,
    ("Se", 0, 2): 6,
    ("Se", 1, 0): 3,
    ("Se", 1, 1): 5,
    ("Se", -1, 0): 1,
    ("Se", -2, 0): 0,
    ("Cs", 1, 0): 0,
    ("Cs", 0, 0): 1,
    ("Ba", 0, 0): 2,
    ("Ba", 2, 0): 0,
    ("Bi", 0, 0): 3,
    ("Bi", 3, 0): 0,
    ("Br", 0, 0): 1,
    ("Br", -1, 0): 0,
    ("Br", 2, 0): 3,
    ("Zr", 4, 0): 0,
    ("Kr", 0, 0): 0,
    ("Rb", 0, 0): 1,
    ("Rb", 1, 0): 0,
    ("Br", -1, 0): 0,
    ("Sr", 0, 0): 2,
    ("Sr", 2, 0): 0,
    ("Ag", 0, 0): 2,
    ("Ag", 1, 0): 0,
    ("Ag", -4, 0): 3,
    ("Te", 0, 0): 2,
    ("Te", 1, 0): 3,
    ("Te", 0, 1): 4,
    ("Te", 0, 2): 6,
    ("Te", -1, 1): 3,
    ("Te", -1, 2): 5,
    ("I", 0, 0): 1,
    ("I", 0, 1): 3,
    ("I", 0, 2): 5,
    ("I", -1, 0): 0,
    ("I", 1, 0): 2,
    ("I", 2, 1): 3,
    ("I", 3, 0): 4,
    ("At", 0, 0): 1,
    ("Au", 0, 0): 0,
    ("Ra", 0, 0): 2,
    ("Ra", 2, 0): 0,
}

BANGS = {
    (sym, chg, val): bangs for (sym, chg, bangs), val in VALENCE.items() if bangs > 0
}

File: rust_amsr/src/valence.rs
use std::collections::HashMap;
use lazy_static::lazy_static;

lazy_static! {
    static ref VALENCE: HashMap<(String, i32, i32), i32> = {
        let mut map = HashMap::new();
        
        // Hydrogen
        map.insert(("H".to_string(), 0, 0), 1);
        map.insert(("H".to_string(), 1, 0), 0);
        map.insert(("H".to_string(), -1, 0), 0);
        
        // Helium
        map.insert(("He".to_string(), 0, 0), 0);
        
        // Lithium
        map.insert(("Li".to_string(), 1, 0), 0);
        map.insert(("Li".to_string(), 0, 0), 1);
        
        // Beryllium
        map.insert(("Be".to_string(), 2, 0), 0);
        
        // Boron
        map.insert(("B".to_string(), 0, 0), 3);
        map.insert(("B".to_string(), -1, 0), 4);
        
        // Carbon
        map.insert(("C".to_string(), 0, 0), 4);
        map.insert(("C".to_string(), -1, 0), 3);
        map.insert(("C".to_string(), 1, 0), 3);
        
        // Nitrogen
        map.insert(("N".to_string(), 0, 0), 3);
        map.insert(("N".to_string(), 1, 0), 4);
        map.insert(("N".to_string(), -1, 0), 2);
        
        // Oxygen
        map.insert(("O".to_string(), 0, 0), 2);
        map.insert(("O".to_string(), -1, 0), 1);
        map.insert(("O".to_string(), -2, 0), 0);
        map.insert(("O".to_string(), 1, 0), 3);
        
        // Fluorine
        map.insert(("F".to_string(), 0, 0), 1);
        map.insert(("F".to_string(), -1, 0), 0);
        map.insert(("F".to_string(), 1, 0), 2);
        
        // Sodium
        map.insert(("Na".to_string(), 0, 0), 1);
        map.insert(("Na".to_string(), 1, 0), 0);
        
        // Magnesium
        map.insert(("Mg".to_string(), 0, 0), 2);
        map.insert(("Mg".to_string(), 0, 1), 6);
        map.insert(("Mg".to_string(), 1, 0), 1);
        map.insert(("Mg".to_string(), 2, 0), 0);
        
        // Aluminum
        map.insert(("Al".to_string(), 0, 0), 3);
        map.insert(("Al".to_string(), 3, 0), 0);
        map.insert(("Al".to_string(), -3, 0), 6);
        
        // Xenon
        map.insert(("Xe".to_string(), 0, 0), 0);
        
        // Silicon
        map.insert(("Si".to_string(), 0, 0), 4);
        map.insert(("Si".to_string(), -1, 0), 5);
        map.insert(("Si".to_string(), 4, 0), 0);
        
        // Phosphorus
        map.insert(("P".to_string(), 0, 0), 3);
        map.insert(("P".to_string(), 0, 1), 5);
        map.insert(("P".to_string(), 0, 2), 7);
        map.insert(("P".to_string(), 1, 0), 4);
        map.insert(("P".to_string(), -1, 1), 6);
        
        // Sulfur
        map.insert(("S".to_string(), 0, 0), 2);
        map.insert(("S".to_string(), 0, 1), 4);
        map.insert(("S".to_string(), 0, 2), 6);
        map.insert(("S".to_string(), 1, 0), 3);
        map.insert(("S".to_string(), 1, 1), 5);
        map.insert(("S".to_string(), -1, 0), 1);
        map.insert(("S".to_string(), -1, 1), 3);
        map.insert(("S".to_string(), -1, 2), 5);
        map.insert(("S".to_string(), -2, 0), 0);
        
        // Chlorine
        map.insert(("Cl".to_string(), 0, 0), 1);
        map.insert(("Cl".to_string(), -1, 0), 0);
        map.insert(("Cl".to_string(), 1, 0), 2);
        map.insert(("Cl".to_string(), 2, 0), 3);
        map.insert(("Cl".to_string(), 3, 0), 4);
        
        // Potassium
        map.insert(("K".to_string(), 0, 0), 1);
        map.insert(("K".to_string(), 1, 0), 0);
        
        // Calcium
        map.insert(("Ca".to_string(), 0, 0), 2);
        map.insert(("Ca".to_string(), 2, 0), 0);
        
        // Iron
        map.insert(("Fe".to_string(), 3, 0), 0);
        
        // Zinc
        map.insert(("Zn".to_string(), 0, 0), 2);
        map.insert(("Zn".to_string(), 1, 0), 1);
        map.insert(("Zn".to_string(), 2, 0), 1);
        map.insert(("Zn".to_string(), -2, 0), 2);
        map.insert(("Zn".to_string(), -2, 1), 4);
        
        // Arsenic
        map.insert(("As".to_string(), 0, 0), 3);
        map.insert(("As".to_string(), 0, 1), 5);
        map.insert(("As".to_string(), 0, 2), 7);
        map.insert(("As".to_string(), 1, 0), 4);
        map.insert(("As".to_string(), -1, 1), 6);
        
        // Selenium
        map.insert(("Se".to_string(), 0, 0), 2);
        map.insert(("Se".to_string(), 0, 1), 4);
        map.insert(("Se".to_string(), 0, 2), 6);
        map.insert(("Se".to_string(), 1, 0), 3);
        map.insert(("Se".to_string(), 1, 1), 5);
        map.insert(("Se".to_string(), -1, 0), 1);
        map.insert(("Se".to_string(), -2, 0), 0);
        
        // Cesium
        map.insert(("Cs".to_string(), 1, 0), 0);
        map.insert(("Cs".to_string(), 0, 0), 1);
        
        // Barium
        map.insert(("Ba".to_string(), 0, 0), 2);
        map.insert(("Ba".to_string(), 2, 0), 0);
        
        // Bismuth
        map.insert(("Bi".to_string(), 0, 0), 3);
        map.insert(("Bi".to_string(), 3, 0), 0);
        
        // Bromine
        map.insert(("Br".to_string(), 0, 0), 1);
        map.insert(("Br".to_string(), -1, 0), 0);
        map.insert(("Br".to_string(), 2, 0), 3);
        
        // Zirconium
        map.insert(("Zr".to_string(), 4, 0), 0);
        
        // Krypton
        map.insert(("Kr".to_string(), 0, 0), 0);
        
        // Rubidium
        map.insert(("Rb".to_string(), 0, 0), 1);
        map.insert(("Rb".to_string(), 1, 0), 0);
        
        // Strontium
        map.insert(("Sr".to_string(), 0, 0), 2);
        map.insert(("Sr".to_string(), 2, 0), 0);
        
        // Silver
        map.insert(("Ag".to_string(), 0, 0), 2);
        map.insert(("Ag".to_string(), 1, 0), 0);
        map.insert(("Ag".to_string(), -4, 0), 3);
        
        // Tellurium
        map.insert(("Te".to_string(), 0, 0), 2);
        map.insert(("Te".to_string(), 1, 0), 3);
        map.insert(("Te".to_string(), 0, 1), 4);
        map.insert(("Te".to_string(), 0, 2), 6);
        map.insert(("Te".to_string(), -1, 1), 3);
        map.insert(("Te".to_string(), -1, 2), 5);
        
        // Iodine
        map.insert(("I".to_string(), 0, 0), 1);
        map.insert(("I".to_string(), 0, 1), 3);
        map.insert(("I".to_string(), 0, 2), 5);
        map.insert(("I".to_string(), -1, 0), 0);
        map.insert(("I".to_string(), 1, 0), 2);
        map.insert(("I".to_string(), 2, 1), 3);
        map.insert(("I".to_string(), 3, 0), 4);
        
        // Astatine
        map.insert(("At".to_string(), 0, 0), 1);
        
        // Gold
        map.insert(("Au".to_string(), 0, 0), 0);
        
        // Radium
        map.insert(("Ra".to_string(), 0, 0), 2);
        map.insert(("Ra".to_string(), 2, 0), 0);
        
        map
    };
    
    static ref BANGS: HashMap<(String, i32, i32), i32> = {
        let mut map = HashMap::new();
        
        // Generate BANGS table from VALENCE table
        for ((sym, chg, bangs), val) in VALENCE.iter() {
            if *bangs > 0 {
                map.insert((sym.clone(), *chg, *val), *bangs);
            }
        }
        
        map
    };
}

pub fn get_valence(atom_symbol: &str, charge: i32, bangs: i32) -> Option<i32> {
    VALENCE.get(&(atom_symbol.to_string(), charge, bangs)).copied()
}

pub fn get_bangs(atom_symbol: &str, charge: i32, total_valence: i32) -> i32 {
    BANGS.get(&(atom_symbol.to_string(), charge, total_valence)).copied().unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_carbon_valence() {
        assert_eq!(get_valence("C", 0, 0), Some(4));
        assert_eq!(get_valence("C", 1, 0), Some(3));
        assert_eq!(get_valence("C", -1, 0), Some(3));
    }

    #[test]
    fn test_oxygen_valence() {
        assert_eq!(get_valence("O", 0, 0), Some(2));
        assert_eq!(get_valence("O", 1, 0), Some(3));
        assert_eq!(get_valence("O", -1, 0), Some(1));
        assert_eq!(get_valence("O", -2, 0), Some(0));
    }

    #[test]
    fn test_sulfur_bangs() {
        assert_eq!(get_bangs("S", 0, 4), 1);
        assert_eq!(get_bangs("S", 0, 6), 2);
        assert_eq!(get_bangs("S", 1, 5), 1);
    }

    #[test]
    fn test_phosphorus_bangs() {
        assert_eq!(get_bangs("P", 0, 5), 1);
        assert_eq!(get_bangs("P", 0, 7), 2);
        assert_eq!(get_bangs("P", -1, 6), 1);
    }
} 
